broker/features.go

```diff
@@ -1,26 +1,41 @@
 package broker
 
 import (
+	"fmt"
+
 	"github.com/dingotiles/dingo-postgresql-broker/broker/structs"
 	"github.com/frodenas/brokerapi"
+	"github.com/mitchellh/mapstructure"
 )
 
-func (bkr *Broker) clusterFeaturesFromProvisionDetails(details brokerapi.ProvisionDetails) structs.ClusterFeatures {
-	targetNodeCount := defaultNodeCount
-	if rawNodeCount := details.Parameters["node-count"]; rawNodeCount != nil {
-		targetNodeCount = int(rawNodeCount.(float64))
+func (bkr *Broker) clusterFeaturesFromProvisionDetails(details brokerapi.ProvisionDetails) (features structs.ClusterFeatures, err error) {
+	return bkr.clusterFeaturesFromParameters(details.Parameters)
+}
+
+func (bkr *Broker) clusterFeaturesFromUpdateDetails(details brokerapi.UpdateDetails) (features structs.ClusterFeatures, err error) {
+	return bkr.clusterFeaturesFromParameters(details.Parameters)
+}
+
+func (bkr *Broker) clusterFeaturesFromParameters(params map[string]interface{}) (features structs.ClusterFeatures, err error) {
+	err = mapstructure.Decode(params, &features)
+	if err != nil {
+		return
 	}
-	return structs.ClusterFeatures{
-		NodeCount: targetNodeCount,
+	if features.NodeCount < 2 {
+		features.NodeCount = 2
 	}
+
+	err = bkr.verifyClusterFeatures(features)
+	return
 }
 
-func (bkr *Broker) clusterFeaturesFromUpdateDetails(details brokerapi.UpdateDetails) structs.ClusterFeatures {
-	targetNodeCount := defaultNodeCount
-	if rawNodeCount := details.Parameters["node-count"]; rawNodeCount != nil {
-		targetNodeCount = int(rawNodeCount.(float64))
+func (bkr *Broker) verifyClusterFeatures(features structs.ClusterFeatures) (err error) {
+	availableCells, err := bkr.scheduler.FilterCellsByGUIDs(features.CellGUIDsForNewNodes)
```

Comment:
Can we extract the verification.
My suggestion is put the logic into the scheduler `bkr.scheduler.VerifyClusterFeaters` and call it in the `assert*Preconditions` method. Every endpoint has one like (here)[https://github.com/dingotiles/dingo-postgresql-broker/blob/master/broker/provision_endpoint.go#L83]